pkg/apis/core/validation/validation.go

```diff
@@ -5328,8 +5329,9 @@ func ValidateResourceRequirements(requirements *core.ResourceRequirements, fldPa
 		// Validate resource quantity.
 		allErrs = append(allErrs, ValidateResourceQuantityValue(string(resourceName), quantity, fldPath)...)
 
-		if helper.IsHugePageResourceName(resourceName) {
-			limContainsHugePages = true
+		limContainsHugePages, err = validateResourceQuantityHugePageValue(resourceName, quantity)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, quantity.String(), err.Error()))
 		}
```

Comment:
Tightening validation here can invalidate data that was persisted in previous releases. This validation needs to tolerate updates that do not change existing bad data.

This will likely look like:
1. add `AllowIndivisibleHugepagesValues bool` to `PodValidationOptions`
2. set that to `true` in `GetValidationOptionsFromPodSpec` if the existing pod spec has indivisible hugepages values
3. plumb pod validation options to this method
4. allow indivisible hugepages values if indicated in the options

Add unit tests covering all of these scenarios:
1. run create validation on a pod with indivisible hugepages values → should fail
2. run create validation on a deployment with indivisible hugepages values → should fail
3. run update validation on an existing pod with indivisible hugepages values to a new pod with indivisible hugepages values → should succeed
4. run update validation on an existing deployment with indivisible hugepages values to a new deployment with indivisible hugepages values → should succeed
5. run update validation on an existing deployment with **divisible** hugepages values to a new deployment with indivisible hugepages values → should fail


cc @derekwaynecarr for guidance

Reply:
@liggitt above flow looks good.

it maps what we had to do for AllowDownwardAPIForHugePages and not break existing data.

Reply:
@lala123912 a good pr to see similar flow above for reference is https://github.com/kubernetes/kubernetes/pull/86102

Reply:
thanks for the advices.
I have updated the code , pls reivew it again. @liggitt @derekwaynecarr